### PR TITLE
feat(proxy): push all branches when creating a project

### PR DESCRIPTION
**Pull request by rudolfs**
Closes: https://github.com/radicle-dev/radicle-upstream/issues/1249.

### CODE REVIEW

File: proxy/coco/src/project/create/validation.rs [hunk 1]
```diff
@@ -266,16 +267,36 @@ impl Repository {
         log::debug!("Creating rad remote");
         let mut git_remote = Self::existing_remote(repo, &url)?
             .map_or_else(|| Remote::rad_remote(url, None).create(repo), Ok)?;
-        Self::push_default(&mut git_remote, default_branch)?;
+        // Self::push_default(&mut git_remote, default_branch)?;
```
Comment (geigerzaehler) [inline proxy/coco/src/project/create/validation.rs:270]: Leftover comment

File: proxy/coco/src/project/create/validation.rs [hunk 2]
```diff
@@ -266,16 +267,36 @@ impl Repository {
...
-    /// Push the default branch to the provided remote.
-    fn push_default(remote: &mut git2::Remote, default_branch: &OneLevel) -> Result<(), Error> {
-        let default: FlatRef<RefLike, _> =
-            FlatRef::head(PhantomData, None, default_branch.clone().into());
-        log::debug!("Pushing default branch '{}'", default);
-        remote.push(&[default.to_string()], None)?;
+    fn push_branches(repo: &git2::Repository, remote: &mut git2::Remote) -> Result<(), Error> {
+        let local_branches = repo
+            .branches(Some(git2::BranchType::Local))
+            .map_err(git::error::Error::from)?
```
Comment (geigerzaehler) [inline proxy/coco/src/project/create/validation.rs:278]: The explicit `.map_err()` is not needed. The `?` operator takes care of this.

```suggestion
            .branches(Some(git2::BranchType::Local))?
```

File: proxy/api/src/http/project.rs [hunk 3]
```diff
@@ -507,7 +507,7 @@ mod test {
             },
             "shareableEntityIdentifier": format!("%{}", meta.id.to_string()),
             "stats": {
-                "branches": 1,
+                "branches": 2,
```
Comment (geigerzaehler) [inline proxy/api/src/http/project.rs:510]: Is this enough of a test?

**Review by geigerzaehler — COMMENTED** [review]
Take my review with a grain of salt since I’m not very familiar with the code

File: proxy/coco/src/project/create/validation.rs [hunk 4]
```diff
@@ -266,16 +267,36 @@ impl Repository {
...
+            .filter_map(|branch_result| {
+                let (branch, _) = branch_result.ok()?;
+                let name = branch.name().ok()?;
+                name.map(String::from)
```
Comment (FintanH) [inline proxy/coco/src/project/create/validation.rs:282]: We could do the prefixing here

File: proxy/coco/src/project/create/validation.rs [hunk 5]
```diff
@@ -266,16 +267,36 @@ impl Repository {
...
+            });
+
+        let local_prefixed_branches = local_branches
+            .map(|branch| format!("refs/heads/{}", branch))
+            .collect::<Vec<String>>();
+
+        log::debug!(
+            "Pushing branches {}",
+            local_prefixed_branches
```
Comment (FintanH) [inline proxy/coco/src/project/create/validation.rs:291]: I would just debug the `Vec`, it shouldn't be a huge deal.

**Review by FintanH — CHANGES_REQUESTED** [review]
Ya, this looks like it's on the right track :+1:

File: proxy/coco/src/project/create/validation.rs (hunk 1 above)
Comment (rudolfs) [reply proxy/coco/src/project/create/validation.rs]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/39fb63f27a439a26eb2491fdeaea9a05e33e5493

File: proxy/coco/src/project/create/validation.rs (hunk 2 above)
Comment (rudolfs) [reply proxy/coco/src/project/create/validation.rs]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/f302bf0abec12937206a43538e17417c4572cce1

File: proxy/coco/src/project/create/validation.rs (hunk 4 above)
Comment (rudolfs) [reply proxy/coco/src/project/create/validation.rs]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/68678c06611100d6ea512aadc99b145483100e69

File: proxy/coco/src/project/create/validation.rs (hunk 5 above)
Comment (rudolfs) [reply proxy/coco/src/project/create/validation.rs]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/68678c06611100d6ea512aadc99b145483100e69

File: proxy/coco/src/project/create/validation.rs [hunk 6]
```diff
@@ -31,7 +28,11 @@ pub enum Error {
 
     /// An error occurred in `git2` that we could not handle.
     #[error(transparent)]
-    Git(#[from] git2::Error),
+    Git2(#[from] git2::Error),
+
+    /// An error occurred during a [`radicle_surf::git`] operation.
+    #[error(transparent)]
+    Git(#[from] git::error::Error),
```
Comment (FintanH) [inline proxy/coco/src/project/create/validation.rs:35]: We don't need this error variant now, as far as I can tell. Could we remove it?

**Review by FintanH — CHANGES_REQUESTED** [review]
Looking better!

File: proxy/coco/src/project/create/validation.rs (hunk 6 above)
Comment (rudolfs) [reply proxy/coco/src/project/create/validation.rs]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/4500f4d7bd06e6d6fc39956767300c52e3e293c5

**Review by FintanH — DISMISSED** [review]
Better tests would be nice, but in general I'm happy with this :+1:

**Review by xla — DISMISSED** [review]
🐈 🏸 🆘 🍉

File: proxy/api/src/http/project.rs (hunk 3 above)
Comment (rudolfs) [reply proxy/api/src/http/project.rs:531]: https://github.com/radicle-dev/radicle-upstream/pull/1278/commits/92d6601c9e38f3a7efe28ad6832a7e03f1244a5b

**Review by xla — DISMISSED** [review]
🍭 🅰️ 🈶 🌩

**Review by FintanH — APPROVED** [review]

**Review by xla — APPROVED** [review]
🌯 💊 😎 🍦

**Pull request merged**